In [1]:
# AF1 - Linear Regression: Predicting Monthly Spend
# Import the required libraries
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

In [6]:
# 1. Data loading
data = {
    "customer_id": [1, 2, 3, 4, 5, 6, 7, 8],
    "monthly_visits": [5, 8, 4, 10, 7, 6, 9, 3],
    "time_in_store_min": [20, 35, 15, 40, 25, 22, 38, 12],
    "monthly_spend_usd": [150, 245, 110, 300, 210, 180, 275, 90],
}
df = pd.DataFrame(data)
df

,customer_id,monthly_visits,time_in_store_min,monthly_spend_usd
0,1,5,20,150
1,2,8,35,245
2,3,4,15,110
3,4,10,40,300
4,5,7,25,210
5,6,6,22,180
6,7,9,38,275
7,8,3,12,90


In [ ]:
# 2. Exploratory analysis
# Descriptive statistics and correlation between variables
print("Description:")
print(df.describe())
print()
print("Correlation matrix:")
print(df[["monthly_visits", "time_in_store_min", "monthly_spend_usd"]].corr())

In [14]:
# 3. Model training
# Predictor variables (X) and target variable (y)
X = df[["monthly_visits", "time_in_store_min"]]
y = df["monthly_spend_usd"]

model = LinearRegression()
model.fit(X, y)

w0=model.intercept_
w1=model.coef_[0]
w2=model.coef_[1]

print("W0: ", model.intercept_)
print("W1: ", model.coef_[0])
print("W2: ", model.coef_[1])
print("Coefficients:", dict(zip(X.columns, model.coef_)))
print("Intercept:", model.intercept_)

#manual prediction
new_visits= 8
new_time=25
prediction = w0 + w1 * new_visits + w2 * new_time

print("Manual Prediction: ", prediction)

#Library Prediction
new_customer=[[8,25]]
prediction=  model.predict(pd.DataFrame(new_customer, columns=["monthly_visits", "time_in_store_min"]))
print("Linear Regression Prediction: ", prediction)


W0:  -5.143627696939205
W1:  26.98632714500751
W2:  0.9558454591068744
Coefficients: {'monthly_visits': np.float64(26.98632714500751), 'time_in_store_min': np.float64(0.9558454591068744)}
Intercept: -5.143627696939205
Manual Prediction:  234.64312594079274
Linear Regression Prediction:  [234.64312594]


In [19]:
# 4. Model evaluation
df["predicted_spend_usd"] = model.predict(X).round(2)
mae = mean_absolute_error(y, df["predicted_spend_usd"])

print(f"MAE (Mean Absolute Error): {mae:.2f} USD")
df[["customer_id", "monthly_spend_usd", "predicted_spend_usd"]]

MAE (Mean Absolute Error): 2.52 USD


,customer_id,monthly_spend_usd,predicted_spend_usd
0,1,150,148.90
1,2,245,244.20
2,3,110,117.14
3,4,300,302.95
4,5,210,207.66
5,6,180,177.80
6,7,275,274.06
7,8,90,87.29


**Preguntas:**

**¿Qué representa la matriz de coeficientes que obtuviste?:**
* -Los coeficientes representan cuanto cambia el gasto mensual, cuando cambia cada variable.
* Por ejemplo, por cada visita adicional al mes **W1:**, el gasto se incrementa en : $26.99 y por cada minuto adicional: **W2:**: el gasto aumenta en $0.96. Para medir el efecto de cada variable, se mantienen las otras como constantes.

**¿Qué te dice la métrica de desempeño sobre tu modelo?**
* La MAE **(2.52)**, indica que las predicciones del modelo se desvían en promedio, por aproximadamente $2.52 del gasto mensual real.
*Analizando los datos del dataset, en donde el gasto mensual varía entre 90 y 300, el error es pequeño, menor al 2% del valor promedio. Esto sugiere que el modelo se aproxima a la relación entre visitas, el tiempo en tienda y el gasto mensual.

* Sin embargo, dado que este MAE se calculó sobre los datos que fueron usados para entrenar el modelo (actualmente no existe entrenamiento/testing) y que el dataset utilizado es demasiado pequeño (8 valores), el error bajo refleja que el modelo se ajusta a estos datos específicos, pero podría no garantizar el mismo desempeño con datos de clientes nuevos.

